### PR TITLE
Add compression to protocol

### PR DESCRIPTION
**Pull request by justus-camp**
## Description

To indicate if the op is compressed and what algorithm is being used to compress it, we need to include the algorithm as part of the protocol. This would allow the backend to decompress the op if it wishes. 

## Reviewer Guidance

I'm not sure what other things need to happen as a result of this change. Will this property get round-tripped back by our supported backends or does additional work need to occur for that to happen?

### CODE REVIEW

**Comment by DLehenbauer** [comment]
@milanro, @dstanesc - Just FYI - Early / experimental draft PR.

@dstanesc - I saw that you were experimenting with e2e latency:
image

Looks like LZ4 continues to be the best tradeoff?

**Comment by dstanesc** [comment]
Yes, it looks like for the transfer of the payloads we have in mind today for larger data (chunking considered) and the available network speed  `lz4` adds the least overhead. If overall size is a problem (eg. want to minimize costs), data at rest (ie. persistence) may benefit from additional compression ( ~ 2x). 

2nd level of compression could be regarded as an implementation preference of the FRS infrastructure. 

> This would allow the backend to decompress the op if it wishes.

To @justus-camp comment, would be very interesting to analyze & understand the use-cases mandating backends to decompress container data.

File: common/lib/protocol-definitions/src/protocol.ts [hunk 1]
```diff
@@ -93,6 +102,12 @@ export interface IDocumentMessage {
 
     // Traces related to the packet.
     traces?: ITrace[];
+
```
Comment (vladsud) [inline common/lib/protocol-definitions/src/protocol.ts:105]: I believe you want to make same change for ISequencedDocumentMessage.

**Comment by vladsud** [comment]
@GaryWilber - FYI, can you please point place in Deli that we need to change to ensure new property will go through?

File: common/lib/protocol-definitions/src/protocol.ts [hunk 2]
```diff
@@ -93,6 +102,12 @@ export interface IDocumentMessage {
...
+    /**
+     * The compression algorithm that was used to compress the op.
+     * @experimental Not ready for use
+     */
+    compression?: CompressionAlgorithm;
```
Comment (vladsud) [inline common/lib/protocol-definitions/src/protocol.ts:110]: I'd say - at protocol level, this needs to be a string, i.e. not hard code only possible values. In other words, adding new compression algorithm should not be protocol breaking. I'd move actual list of compression algorithms to runtime. While storage may want to know and alter content, it's not required, and I think we will not expect that to happen (at least not in relay service)

**Comment by GaryWilber** [comment]
> @GaryWilber - FYI, can you please point place in Deli that we need to change to ensure new property will go through?

There's two changes that will need to be made.
1. Update alfred to pass the property to deli - [include the new property in this object](https://github.com/microsoft/FluidFramework/blob/380d06ef4207e2beab63bf5ae274b24e0078749a/server/routerlicious/packages/lambdas/src/alfred/index.ts#L86)
2. Update deli to include the property in sequenced ops - [include the new property in this object](https://github.com/microsoft/FluidFramework/blob/380d06ef4207e2beab63bf5ae274b24e0078749a/server/routerlicious/packages/lambdas/src/deli/lambda.ts#L1186)

File: common/lib/protocol-definitions/src/protocol.ts (hunk 1 above)
Comment (justus-camp) [reply common/lib/protocol-definitions/src/protocol.ts:105]: You're correct, this was opened quickly to start discussion. I'll update the PR now to include the property  ISequencedDocumentMessage along with the changes where Gary has pointed us to

File: common/lib/protocol-definitions/src/protocol.ts (hunk 2 above)
Comment (justus-camp) [reply common/lib/protocol-definitions/src/protocol.ts:110]: Yeah I wasn't sure what strategy we wanted to go for here. Using a string here definitely gives us more flexibility if we want to add more algorithms in the future without a breaking change. I'll change this to a string property and add the compression strings to the runtime.

**Comment by justus-camp** [comment]
Do those changes need to go in a separate PR or can they be included in this one?

**Comment by GaryWilber** [comment]
I think those changes can be included in this PR, though the protocol def updates won't be available in the server packages until there's a version release/bump. So you would need some `as any`'s

File: common/lib/protocol-definitions/src/protocol.ts [hunk 3]
```diff
@@ -165,6 +171,12 @@ export interface ISequencedDocumentMessage {
      * @alpha
      */
     expHash1?: string;
+
+    /**
+     * The compression algorithm that was used to compress this op.
```
Comment (vladsud) [inline common/lib/protocol-definitions/src/protocol.ts:176]: I'd say not "this op" but "contents of this op". It's a bit clearer what is compressed.
Maybe we should call it contentCompression, but I hate all these longs strings - op envelope is already over 200 bytes!

**Review by vladsud — APPROVED** [review]
It probably does not matter, but you could submit this change to main as it's not breaking. That will help with propagating this change to ordering service sooner

**Comment by vladsud** [comment]
One interesting observation: Client can't use compression (newly added property) until it's sure relay service (aka ordering service supports it). There are two ways to ensure we are on correct path:
1. We wait enough time for changes to relay service propagates to 100% production. This might be fine for ODSP, but FRS will have different / slower timeline, so likely we will need to find a way to enable compression only for ODSP workflows (likely - via making it off by default and letting specific applications to enable it).
2. We have mechanism already for relay service to convey to client what features it supports - see IConnect.supportedFeatures. But there are multiple problems with using it (some are easier to solve, some are harder to solve):
 - ops are generated and flushed to PendingStateManager even without connection.
 - there is current way to propagate that info to container runtime at the moment.
 - there is probably exists some chance that if even if we connected to a relay service that supports propagating this property, there might be fall over where client connects in the future to relay service that does not.

I think the easiest way to solve this for now is to use two properties simultaneously - `compression` and `metadata` for now, and eventually remove code that sets `metadata` (though we likely would need to leave code that checks / reads it forever).
# 1 might be preferred, but I do not know how long it takes to get there.

**Comment by justus-camp** [comment]
It's a bit unfortunate, but I think your proposal for using ```metadata``` and ```compression``` property simultaneously seems like the best solution for now. I'm not too familiar with the second option so I'm a bit more inclined to leave it off by default, let applications enable it, and keep writing the ```metadata``` field until we're sure the ordering services will support the new property.

**Comment by vladsud** [comment]
Let's go with that (writing both properties and keeping compression off by default for now). I want to see these changes to relay service happen sooner (in main) and then when we are ready to enable compression feature, re-evaluate if changes propagated fully to prod on ODSP side. If they are, then we may not care about supporting (for reading) metadata property.

File: common/lib/protocol-definitions/api-report/protocol-definitions.api.md [hunk 4]
```diff
@@ -159,11 +159,8 @@ export interface IDocumentAttributes {
 // @public
 export interface IDocumentMessage {
     clientSequenceNumber: number;
-<<<<<<< HEAD
```
Comment (DLehenbauer) [inline common/lib/protocol-definitions/api-report/protocol-definitions.api.md]: @justus-camp - Your doc build issue might be an easy fix?  Looks like there are conflict markers from an unresolved merge.

Comment (justus-camp) [reply common/lib/protocol-definitions/api-report/protocol-definitions.api.md]: I think I finally figured it out. Reading the build more carefully I see:

```
Warning: You have changed the public API signature for this project. Please copy the file "_api-extractor-temp/protocol-definitions.api.md" to "api-report/protocol-definitions.api.md", or perform a local build (which does this automatically). See the Git repo documentation for more info.
```

Just followed those instructions and hopefully CI will pass now 🤞

Comment (justus-camp) [reply common/lib/protocol-definitions/api-report/protocol-definitions.api.md]: But you are correct that I messed up the merge conflict initially. VSCode doesn't save by default when you resolve conflicts in the editor and it gets me pretty often 😄

**Comment by justus-camp** [comment]
From the previous discussion, I've moved these changes to #12135 and am targeting main instead of next

**Pull request closed without merging**